sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -16,76 +16,96 @@
-            # work and the performance benefit will be limited because we
-            # work on a fixed-size subsample of the full data.
-            percentiles = np.linspace(0, 100, num=max_bins + 1)
-            percentiles = percentiles[1:-1]
-            midpoints = np.percentile(col_data, percentiles,
-                                      interpolation='midpoint').astype(X_DTYPE)
-            assert midpoints.shape[0] == max_bins - 1
+    # ignore missing values when computing bin thresholds
+    missing_mask = np.isnan(col_data)
+    if missing_mask.any():
+        col_data = col_data[~missing_mask]
+    col_data = np.ascontiguousarray(col_data, dtype=X_DTYPE)
+    distinct_values = np.unique(col_data)
+    if len(distinct_values) <= max_bins:
+        midpoints = distinct_values[:-1] + distinct_values[1:]
+        midpoints *= .5
+    else:
+        # We sort again the data in this case. We could compute
+        # approximate midpoint percentiles using the output of
+        # np.unique(col_data, return_counts) instead but this is more
+        # work and the performance benefit will be limited because we
+        # work on a fixed-size subsample of the full data.
+        percentiles = np.linspace(0, 100, num=max_bins + 1)
+        percentiles = percentiles[1:-1]
+        midpoints = np.percentile(col_data, percentiles,
+                                  interpolation='midpoint').astype(X_DTYPE)
+        assert midpoints.shape[0] == max_bins - 1
+
+    # We avoid having +inf thresholds: +inf thresholds are only allowed in
+    # a "split on nan" situation.
+    np.clip(midpoints, a_min=None, a_max=ALMOST_INF, out=midpoints)
+    return midpoints
+
+
+def _find_bin_categories(col_data, max_bins, feature_idx):
+    """Extract feature-wise categories from categorical data
+
+    Missing values and negative values (considered missing) are ignored.
+
+    Parameters
+    ----------
+    col_data : array-like, shape (n_features,)
+        The categorical feature to bin.
+    max_bins: int
+        The maximum number of bins to be used for categories, ignoring
+        missing values.
+    feature_idx: int
+        Used for error message if the feature cardinality is greater than
+        max_bins.
 
-        # We avoid having +inf thresholds: +inf thresholds are only allowed in
-        # a "split on nan" situation.
-        np.clip(midpoints, a_min=None, a_max=ALMOST_INF, out=midpoints)
+    Return
+    ------
+    bin: ndarray
+        Map from bin index to categorical value. The size of the array is
+        equal to minimum of `max_bins` and the categories' cardinality,
+        ignoring missing and negative values.
+    """
+    categories = np.unique(col_data)
+
+    # nans and negative values will be considered missing
+    missing = np.isnan(categories)
+    negative = categories < 0
+    both = missing | negative
+    if both.any():
+        categories = categories[~both]
```

Comment:
Is there any reason to treat negative categories as missing now?

I would prefer not doing this: it's just as easy for users to set missing values to `np.nan`

Reply:
This is because `OridinalEncoder` can't passthrough `np.nan` like this yet. (I am kind of waiting for https://github.com/scikit-learn/scikit-learn/pull/17317 to get merged before working on missing values for `OrdinalEncoder`.)

Reply:
Maybe https://github.com/scikit-learn/scikit-learn/pull/18406 will help?

Reply:
Thanks for your review on #18406  !

Would you be happy to drop support for negative values now?

To be honest I'm still not sure how this would help anyway because:

- missing values should be encoded as NaNs anyway, and this estimator properly supports NaNs
- unknown values when binning can only happen in case of early-stopping when some categories are in the validation data but not in the training data. But these categories will come out of the OrdinalEncoder anyway so they won't be negative.

Reply:
looks like this was removed (thanks!)